fbgemm_gpu/src/dram_kv_embedding_cache/feature_evict.h

```diff
@@ -0,0 +1,549 @@
+      block_cursors_[i] = 0;
+      block_nums_snapshot_[i] = 0;
+      shards_finished_.emplace_back(std::make_unique<std::atomic<bool>>(false));
+    }
+  }
+
+  // Initialize shard state.
+  void prepare_evict() {
+    for (int shard_id = 0; shard_id < num_shards_; ++shard_id) {
+      auto rlmap = kv_store_.by(shard_id).rlock();
+      auto* mempool = kv_store_.pool_by(shard_id);
+      block_nums_snapshot_[shard_id] =
+          mempool->get_chunks().size() * mempool->get_blocks_per_chunk();
+      block_cursors_[shard_id] = 0;
+      shards_finished_[shard_id]->store(false);
+    }
+    metrics_.reset();
+  }
+
+  void submit_shard_task(int shard_id) {
+    if (shards_finished_[shard_id]->load()) return;
+    futures_.emplace_back(folly::via(executor_).thenValue(
+        [this, shard_id](auto&&) { process_shard(shard_id); }));
+  }
+
+  void process_shard(int shard_id) {
+    auto start_time = std::chrono::high_resolution_clock::now();
+    std::vector<int64_t> evicted_counts(sub_table_hash_cumsum_.size(), 0);
+    std::vector<int64_t> processed_counts(sub_table_hash_cumsum_.size(), 0);
+    std::vector<float> evict_rates(sub_table_hash_cumsum_.size(), 0.0f);
+
+    auto wlock = kv_store_.by(shard_id).wlock();
+    auto* pool = kv_store_.pool_by(shard_id);
+
+    while (!evict_interrupt_.load() &&
+           block_cursors_[shard_id] < block_nums_snapshot_[shard_id]) {
+      auto* block =
+          pool->template get_block<weight_type>(block_cursors_[shard_id]++);
+      if (block == nullptr) {
+        continue;
+      }
+      int64_t key = FixedBlockPool::get_key(block);
+      int sub_table_id = get_sub_table_id(key);
+      processed_counts[sub_table_id]++;
+      if (evict_block(block, sub_table_id)) {
+        auto it = wlock->find(key);
+        if (it != wlock->end() && block == it->second) {
+          wlock->erase(key);
+          pool->template deallocate_t<weight_type>(block);
+          evicted_counts[sub_table_id]++;
+        }
+      }
+    }
+
+    // Check whether the loop ends normally.
+    if (block_cursors_[shard_id] >= block_nums_snapshot_[shard_id]) {
+      shards_finished_[shard_id]->store(true);
+    }
+
+    auto end_time = std::chrono::high_resolution_clock::now();
+    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(
+        end_time - start_time);
+    {
+      std::lock_guard<std::mutex> lock(metric_mtx_);
+      metrics_.exec_duration_ms += duration.count();
+      for (size_t i = 0; i < evicted_counts.size(); ++i) {
+        metrics_.evicted_counts[i] += evicted_counts[i];
+        metrics_.processed_counts[i] += processed_counts[i];
+      }
+    }
+
+    for (size_t i = 0; i < evicted_counts.size(); ++i) {
+      evict_rates[i] = processed_counts[i] > 0
+                           ? (evicted_counts[i] * 100.0f) / processed_counts[i]
+                           : 0.0f;
+    }
+    DLOG(INFO) << fmt::format(
+        "Shard {} completed: \n"
```

Comment:
this is just one time execution is finished, not the whole shard finished right?
as we'll frequently call the process in every batch, log every time might be too many. like DLOG_EVERY_N(INFO, 1000) can sample logs